src/Analysis/Engine/Impl/DependencyResolution/PathResolverSnapshot.Node.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using Microsoft.PythonTools.Analysis.Infrastructure;
+
+namespace Microsoft.PythonTools.Analysis.DependencyResolution {
+    internal partial struct PathResolverSnapshot {
+        [DebuggerDisplay("{" + nameof(DebuggerDisplay) + ",nq}")]
+        private class Node {
+            private readonly Node[] _children;
+            public readonly string Name;
+            public readonly string ModulePath;
+            public bool IsModule => ModulePath != null;
+            public int ChildrenCount => _children.Length;
+
+            private Node(string name, Node[] children, string modulePath) {
+                Name = name;
+                _children = children;
+                ModulePath = modulePath;
+            }
+
+            public static Node CreateRoot(string path) =>
+                new Node(PathUtils.NormalizePath(path), Array.Empty<Node>(), null);
+
+            public static Node CreateModule(string name, string modulePath)
+                => new Node(name, Array.Empty<Node>(), modulePath);
+
+            public Node(string name, Node child) {
+                Name = name;
+                _children = new[] {child};
+                ModulePath = null;
+            }
+
+            public Node ToModule(string modulePath) => new Node(Name, _children, modulePath);
+            public Node ToPackage() => new Node(Name, _children, null);
+            public Node AddChild(Node child) => new Node(Name, _children.ImmutableAdd(child), ModulePath);
+            public Node ReplaceChildAt(Node child, int index) => new Node(Name, _children.ImmutableReplaceAt(child, index), ModulePath);
```

Comment:
Looks like IList, no?

Reply:
No, it's all immutable, so we can't use `IList` interface.